pkg/backend/filestate/backend.go

```diff
@@ -190,22 +243,65 @@ func New(ctx context.Context, d diag.Sink, originalURL string, project *workspac
 		bucket:      wbucket,
 		lockID:      lockID.String(),
 		gzip:        gzipCompression,
-		store:       newLegacyReferenceStore(wbucket),
 	}
 	backend.currentProject.Store(project)
 
-	// Read the Pulumi state metadata
-	// and ensure that it is compatible with this version of the CLI.
+	// Read the Pulumi state metadata, if any.
+	// This will tell us which store to use.
 	meta, err := ensurePulumiMeta(ctx, wbucket)
 	if err != nil {
 		return nil, err
 	}
-	if meta.Version != 0 {
+
+	// projectMode tracks whether the current state supports project-scoped stacks.
+	// Historically, the filestate backend did not support this.
+	// To avoid breaking old stacks, we use legacy mode for existing states.
+	// We use project mode only if one of the following is true:
+	//
+	//  - The state has a single .pulumi/Pulumi.yaml file
```

Comment:
Should we use a name other than `Pulumi.yaml` so it isn't confused with a project file? I know it doesn't really matter, but I think a different name will help avoid any confusion. `meta.yaml` or `version.yaml` or something else like that?

Reply:
meta.yaml or a plain text version file would have been my initial choice but unfortunately, we can no longer change this.
We've already shipped the change to start writing these .pulumi/Pulumi.yaml files with version 0 two releases ago.

I mean, we _could_ still change this if we really wanted to; it won't break anything. It'll just leave an extraneous unused file if you initialized your bucket with version 3.59 or higher. The file would be ignored by other versions.
If that's acceptable, we can change this.

Reply:
> I mean, we _could_ still change this if we really wanted to; it won't break anything. It'll just leave an extraneous unused file if you initialized your bucket with version 3.59 or higher. The file would be ignored by other versions.
> If that's acceptable, we can change this.

I'd still lean towards changing the filename, as it'll reduce confusion long term.

Reply:
Okay, I'll make that change; probably a good time to also avoid writing the file for legacy layouts so as not to break tight permissioning.

Reply:
Both done; rebasing the stack on top of these:

https://github.com/pulumi/pulumi/pull/12538
https://github.com/pulumi/pulumi/pull/12537